addons/mrp/models/mrp_workorder.py

```diff
@@ -812,15 +812,8 @@ def _update_finished_move(self):
 
     def _check_sn_uniqueness(self):
         """ Alert the user if the serial number as already been produced """
-        if self.product_tracking == 'serial' and self.finished_lot_id:
-            sml = self.env['stock.move.line'].search_count([
-                ('lot_id', '=', self.finished_lot_id.id),
-                ('location_id.usage', '=', 'production'),
-                ('qty_done', '=', 1),
-                ('state', '=', 'done')
-            ])
-            if sml:
-                raise UserError(_('This serial number for product %s has already been produced', self.product_id.name))
+        if self.product_tracking == 'serial' and self.finished_lot_id and self.production_id._is_finished_sn_already_produced(self.product_id, self.finished_lot_id):
+            raise UserError(_('This serial number for product %s has already been produced', self.product_id.name))
```

Comment:
This is a good, I'm just not sure why there isn't already a check in this method for byproducts as well :thinking: Maybe we should add it in? What do you think?

Reply:
I agree
But then, in the same idea, we should also check the uniqueness of the components. To do so, I would extract the components part of `MrpProduction._check_sn_uniqueness` and use it from `MrpWorkorder` (because, again, it would be a shame not to recycle the existing code) 
https://github.com/odoo/odoo/blob/2a03370b946816f4181d219373c9bd3a23d04f35/addons/mrp/models/mrp_production.py#L1740-L1763
Ok with that ?

Reply:
Agreed and OK with it. I tested this out a little more and discovered that this code removes some use case checks :cry: E.g. if a SN exists from a receipt or inventory adjustment. I think you can fix this by reusing the check you have for it in `_onchange_lot_producing` 

https://github.com/odoo/odoo/blob/2a03370b946816f4181d219373c9bd3a23d04f35/addons/mrp/models/mrp_production.py#L691-L696

The same issue occurs in the `_check_sn_uniqueness` of `mrp.production`, but the `check_quantity` constraint check of `stock.quant` catches it.